department-of-veterans-affairs/vets-api
# 62244 - Properly Gate and Connect the AcceptableVerifiedCredentialAdoptionService

## Summary

**Changes Made:**
This PR does the following:
- Changes the name of the already present `reactivation_experiment` Flipper toggle to `reactivation_experiment_rate_limiting` in `app/services/acceptable_verified_credential_adoption_service.rb`
- Adds a new Flipper toggle to `app/services/login/after_login_actions.rb` called `reactivation_experiment_initial_gate` that will control the basic "on/off" for users's being checked for eligibility to be sent a credential adoption related email.
- Updates the `config/features.yml` to align with the previous two points
- Updates the tests for both services to reflect the new and modified feature toggles


**What do these changes solve/make possible?**
These changes enable us to both gate and rate limit the number of users going through the AcceptableVerifiedCredentialAdoptionService flow.  


**Why are you using two Flipper toggles?**
We only want user eligibility checked for the time we have the experiment "running" (outlined below).
We want to restrict the total number of emails being sent out to 5% of daily logged in users who also meet the other eligibility criteria (outlined further below).

More specifically:
- The initial gate feature toggle will give us the ability to easily turn the entire feature on or off to prevent the service from being called and evaluating other eligibility criteria after every time a user logs in.
- The rate limit feature toggle will give us the ability to restrict the percentage of users who meet the other eligibility criteria from receiving the email during this experiment.

Example:
During the experiment window, if 1000 users login daily we want them all to go through the eligibility check within the service (the initial gate toggle).  If 100 users are otherwise eligible, we only want a certain percentage of those folks to actually be sent the email (the rate limit toggle).

Note:  Currently, an eligible user is a veteran user who meets the following criteria: 
- user has both a legacy (DSLogon) and a modern (Login.gov) credential 
- user has most recently logged in with the legacy credential
- user has not triggered a previous Credential Adoption email send in the last seven (7) days


**Team Info:**
This work is being done by the Login.gov Adoption Team which reports to John Rahaghi, Product Owner with OCTO on the VA side.


**Flipper Info:**
`reactivation_experiment_initial_gate` and `reactivation_experiment_rate_limit`
- this toggle will be enabled for one (1) week, starting on/around 20JULY2023
- this toggle will be turned off once the one (1) week time limit is hit OR if we reach 5% of our eligible population, which ever happens first.
- success criteria is based on the percentage of our total eligible veteran user base


## Related issue(s)
- [Connect Credential Adoption Service for Production Usage#62244](https://app.zenhub.com/workspaces/logingov-adoption-team-632280b31e745000136a61fc/issues/gh/department-of-veterans-affairs/va.gov-team/62244)


## Testing done

**Behavior Before Change**
Previously, there was no call to the AcceptableVerifiedCredentialAdoptionService from within the AfterLoginActions service so it was unused code.  

**Verification Steps**
As the VA Notify portion of the flow does not work in a local environment, the process for verifying the behavior was as follows:
- create local users with mocked credentials and attributes that would enable them to be considered eligible
- create local users with mocked credentials and attributes that would NOT enable them to be considered eligible
- temporarily adjust logging methods to log to the console versus StatsD
- simulate logging in with each user, observing that the appropriate log entries and/or database records were or were not created based on if the user was actually eligible or not.

This method of verification was very time consuming and required several mocks and temp changes to logging methods due to our current inability to "pre-test/QA" in a staging environment.  As such I do not recommend it.

The tests for both the AfterLoginAction service and the AcceptableVerifiedCredentialAdoptionService test these behaviors as well, albeit in a limited piecemeal fashion.

## What areas of the site does it impact?
This PR makes no changes to the UI of the site itself.  It enables access to a previously created and tested code path, with that access being controlled by the new/update Flipper toggles.

## Acceptance criteria


## Additional Notes

This feature will be short-lived and will be removed once the Login.gov Adoption Team has fully transitioned to a new GovDelivery instance.  As such, we're trying to keep this service very simple and as decoupled as possible.
